src/Npgsql/Replication/NpgsqlReplicationConnection.cs

```diff
@@ -0,0 +1,918 @@
+                    ? ParseByteaHex(bytes.Slice(2))
+                    : ParseByteaEscape(bytes);
+
+                byte[] ParseByteaHex(ReadOnlySpan<byte> inBytes)
+                {
+                    var outBytes = new byte[inBytes.Length / 2];
+                    for (var i = 0; i < inBytes.Length; i++)
+                    {
+                        var v1 = inBytes[i++];
+                        var v2 = inBytes[i];
+                        outBytes[i / 2] =
+                            (byte)(((v1 - (v1 < 0x3A ? 0x30 : 87)) << 4) | (v2 - (v2 < 0x3A ? 0x30 : 87)));
+                    }
+
+                    return outBytes;
+                }
+
+                byte[] ParseByteaEscape(ReadOnlySpan<byte> inBytes)
+                {
+                    var result = new MemoryStream(new byte[inBytes.Length]);
+                    for (var tp = 0; tp < inBytes.Length;)
+                    {
+                        var c1 = inBytes[tp];
+                        if (c1 != '\\')
+                        {
+                            // Don't validate whether c1 >= 0x20 && c1 <= 0x7e here
+                            // TIMELINE_HISTORY currently (2020-09-13) sends raw
+                            // bytes instead of bytea for the content value.
+                            result.WriteByte(c1);
+                            tp++;
+                            continue;
+                        }
+
+                        var c2 = inBytes[tp + 1];
+                        if (c2 == '\\')
+                        {
+                            result.WriteByte(c2);
+                            tp += 2;
+                            continue;
+                        }
+
+                        var c3 = inBytes[tp + 2];
+                        var c4 = inBytes[tp + 3];
+                        if (c2 >= '0' && c2 <= '3' &&
+                            c3 >= '0' && c3 <= '7' &&
+                            c4 >= '0' && c4 <= '7')
+                        {
+                            c2 <<= 3;
+                            c2 += c3;
+                            c2 <<= 3;
+                            result.WriteByte((byte)(c2 + c4));
+
+                            tp += 4;
+                            continue;
+                        }
+
+                        throw new FormatException("Invalid syntax for type bytea");
+                    }
+
+                    return result.ToArray();
+                }
+            }
+        }
+
+        void SetTimeouts(TimeSpan readTimeout, TimeSpan writeTimeout)
+        {
+            var connector = Connector;
+            connector.UserTimeout = readTimeout > TimeSpan.Zero ? (int)readTimeout.TotalMilliseconds : 0;
+            connector.WriteBuffer.Timeout = writeTimeout;
+        }
+
+        void Cancel()
+        {
+            using var stateResetter = EnsureAndSetState(ReplicationConnectionState.Streaming, ReplicationConnectionState.Cancelling);
```

Comment:
The cancelling state was meant to disable the timers when cancelling.
I haven't reviewed how/if you solve this now but I'll have a look.

Reply:
I agree that there probably was a bug there, though.

Reply:
I haven't yet looked closely at how timers work - it's possible there's a problem there. But the inherently concurrent nature of cancellation makes it a bad idea to change state...

Reply:
> I haven't yet looked closely at how timers work - it's possible there's a problem there.

There is a problem:
https://github.com/npgsql/npgsql/blob/53e75ff3966b21a71afad730c3da42918b0b9d1f/src/Npgsql/Replication/NpgsqlReplicationConnection.cs#L597

Here we'd re-enable the timers we've disabled in the cancel method in a potential race condition.

> But the inherently concurrent nature of cancellation makes it a bad idea to change state...

I don't necessarily agree. There's quite a lot going on concurrently in NpgsqlReplicationConnection.
We've got two timers, the public `SendStatusUpdate()` API that might get called from a thread that's separate from the thread that's enumerating the messages and cancellation/disposal which might happen at any point.

That's why I tried to make the state pretty concurrency safe (only access it via `Interlocked` and check it at the beginning of every non-private method.)
The problem you've found probably was an oversight but we could probably solve it by checking the state ( for Cancelling) before resetting it.

Anyways. Perhaps there's also an alternative solution for the Problem I've pointed out above.

I'll let you think about it before changing anything.

Reply:
After thinking about it I came to the conclusion that there was still a narrow chance for a race condition in my code (if we happen to cancel in the moment right after the timer thread has mover past the `if (State == ReplicationConnectionState.Streaming)` I mentioned above.
We probably need a separate lock object to avoid re-enabling the timers.

Reply:
>  The problem you've found probably was an oversight but we could probably solve it by checking the state ( for Cancelling) before resetting it.

Note that the problem was the opposite - Cancel was resetting the state back to Streaming once cancellation was finished, but the main thread had already reset it back to Idle because from its point of view, Streaming was complete.

I admit that the timers/concurrency is probably the one area I haven't reviewed yet in your PR (was leaving the best for last :wink:), so I don't have anything concrete to say here yet - I will try to dive into this final bit in the next few days. It's generally better to avoid mutating state concurrently when possible - my inspiration here would probably be the existing keepalive stuff on NpgsqlConnector.

Reply:
Cancellation no longer changes state, and after https://github.com/Brar/npgsql/pull/10 doesn't touch the timers either - so everything should be safe.